state-chain/pallets/cf-asset-balances/src/lib.rs

```diff
@@ -0,0 +1,312 @@
+	/// anything.
+	///
+	/// Note that we refund to accounts atomically (we never partially refund an account), whereas
+	/// refunds to vaults or aggkeys can be made incrementally.
+	fn reconcile(
+		chain: ForeignChain,
+		owner: &ExternalOwner,
+		amount_owed: &mut AssetAmount,
+		available: &mut AssetAmount,
+	) -> Result<(), DispatchError> {
+		if *amount_owed > *available {
+			Self::deposit_event(Event::VaultDeficitDetected {
+				chain,
+				amount_owed: *amount_owed,
+				available: *available,
+			});
+		}
+		let amount_reconciled = match chain {
+			ForeignChain::Ethereum | ForeignChain::Arbitrum => match owner {
+				ExternalOwner::Account(address) =>
+					if *amount_owed > *available {
+						0
+					} else {
+						Self::refund_via_egress(chain, address.clone(), *amount_owed)?;
+						*amount_owed
+					},
+				other => {
+					log_or_panic!(
+						"{:?} Liabilities are not supported for chain {:?}.",
+						other,
+						chain
+					);
+					0
+				},
+			},
+			ForeignChain::Polkadot => {
+				match owner {
+					ExternalOwner::AggKey => {
+						let address = ForeignChainAddress::Dot(
+							T::PolkadotKeyProvider::active_epoch_key()
+								// TODO: make this more defensive
+								.expect("Key must exist otherwise we would have nothing to refund.")
+								.key,
+						);
+						let refund_amount = core::cmp::min(*amount_owed, *available);
+						Self::refund_via_egress(chain, address, refund_amount)?;
+						refund_amount
+					},
+					other => {
+						log_or_panic!(
+							"{:?} Liabilities are not supported for chain {:?}.",
+							other,
+							chain
+						);
+						0
+					},
+				}
+			},
+			ForeignChain::Bitcoin | ForeignChain::Solana => match owner {
+				ExternalOwner::Vault => core::cmp::min(*amount_owed, *available),
+				other => {
+					log_or_panic!(
+						"{:?} Liabilities are not supported for chain {:?}.",
+						other,
+						chain
+					);
+					0
+				},
+			},
+		};
+
+		available.defensive_saturating_reduce(amount_reconciled);
+		amount_owed.defensive_saturating_reduce(amount_reconciled);
+
+		Ok(())
+	}
+
+	pub fn trigger_reconciliation() {
+		if !T::SafeMode::get().reconciliation_enabled {
+			log::info!("Reconciliation is disabled. Skipping reconciliation.");
+			return;
+		}
+
+		for chain in ForeignChain::iter() {
+			let owed_assets = Liabilities::<T>::take(chain.gas_asset());
+			WithheldAssets::<T>::mutate(chain.gas_asset(), |total_withheld| {
+				let mut owed_assets = owed_assets.into_iter().collect::<Vec<_>>();
+				owed_assets.sort_by_key(|(_, amount)| core::cmp::Reverse(*amount));
+
+				for (destination, amount) in &mut owed_assets {
+					debug_assert!(*amount > 0);
+					match Self::reconcile(chain, destination, amount, total_withheld) {
+						Ok(_) =>
+							if *total_withheld == 0 {
+								break;
+							},
+						Err(_) => break,
+					}
+				}
+
+				owed_assets.retain(|(_, amount)| *amount > 0);
```

Comment:
That has been added by Dan, but I think the idea is to clean up the map entries after we refunded. I don't think we want to take over the old empty entries over to the next refund round and then clean it up right before we start the new refund. Apart from that, I don't think it would make a difference in terms of removing the debug_assert. With the current implementation it should also actually never happen, and if it does, we know there is a logical problem.